src/callbacks_stage/positivity_shallow_water_dg2d.jl

```diff
@@ -0,0 +1,90 @@
+# By default, Julia/LLVM does not use fused multiply-add operations (FMAs).
+# Since these FMAs can increase the performance of many numerical algorithms,
+# we need to opt-in explicitly.
+# See https://ranocha.de/blog/Optimizing_EC_Trixi for further details.
+@muladd begin
+
+
+function limiter_shallow_water!(u, threshold::Real, variable,
+                                mesh::AbstractMesh{2},
+                                equations::ShallowWaterEquations2D, dg::DGSEM, cache)
+  @unpack weights = dg.basis
+
+  @threaded for element in eachelement(dg, cache)
+    # determine minimum value
+    value_min = typemax(eltype(u))
+    for j in eachnode(dg), i in eachnode(dg)
+      u_node = get_node_vars(u, equations, dg, i, j, element)
+      value_min = min(value_min, variable(u_node, equations))
+    end
+
+    # detect if limiting is necessary
+    value_min < threshold || continue
+
+    # compute mean value
+    u_mean = zero(get_node_vars(u, equations, dg, 1, 1, element))
+    for j in eachnode(dg), i in eachnode(dg)
+      u_node = get_node_vars(u, equations, dg, i, j, element)
+      u_mean += u_node * weights[i] * weights[j]
+    end
+    # note that the reference element is [-1,1]^ndims(dg), thus the weights sum to 2
+    u_mean = u_mean / 2^ndims(mesh)
+
+    # We compute the value directly with the mean values, as we assume that
+    # Jensen's inequality holds (e.g. pressure for compressible Euler equations).
+    value_mean = variable(u_mean, equations)
+    theta = (value_mean - threshold) / (value_mean - value_min)
+    for j in eachnode(dg), i in eachnode(dg)
+      u_node = get_node_vars(u, equations, dg, i, j, element)
+      
+      # Cut off velocity in case that the water height is smaller than the threshold
+
+      h_node, h_v1_node, h_v2_node, b_node = u_node
+      h_mean, h_v1_mean, h_v2_mean, _ = u_mean # b_mean is not used as it must not be overwritten
+
+      if h_node <= threshold
+        h_v1_node = 0
+        h_v2_node = 0
+        h_v1_mean = 0
+        h_v2_mean = 0
+      end
```

Comment:
Type instability

Reply:
Should this be resolved by (e.g.) `h_v1_node = zero(eltype(u))`?

Reply:
I would also need changed in the "safety" loop. Basically you should avoid setting actual zeros like this. Another reason is that `0` is an `Int` which would change type when it gets saved elsewhere as a `Float`.

Reply:
Done